### PR TITLE
Bias PPE towards term being printed

### PR DESCRIPTION
**Pull request by ChrisPenner**
Built on #3194 

I've got a bit of a longer term plan for PPE's [here](https://github.com/unisonweb/unison/issues/3256), but it's dangerous to make too many changes to how names are handled at once. This adds biasing, which we've wanted for a while.

It's long been an issue in unison where, when viewing a definition, some of the names might be unexpectedly from some other definition, and sometimes even the name of the thing you asked for gets replaced with a different name for that hash;

E.g. if I type `view trunk.List.map` it might print everything using the name `a1.List.map` instead, and even though those ARE valid names for the hashes, it's not what I asked for.

This fixes that by:

* When printing something the user requested, prefer names with the _longest common name-segment prefix_ with the name being printed.

E.g. if you're printing `base.trunk.List.foo`, which mentions `base.trunk.List.bar`, `lengthCommonPrefix(base.trunk.List.foo, base.trunk.List.bar) == 3` whereas `lengthCommonPrefix(base.trunk.List.foo, base.latest.List.bar) == 1`, so `base.trunk.List.bar` is preferred in the pretty printer, which is nice, because it's closest to what we're actually trying to print.

If two names have the same common-prefix length, we fall back on the current name-selection strategies.

## viewing `trunk.Seq` on `distributed` before:

image

## viewing `trunk.Seq` on `distributed` after:

image

## Implementation notes

* Alters PrettyPrintEnv so it keeps all names matching the ref until a single name needs to be selected, this allows us to just 'fmap' different biases over the list of names, rather than building a whole new ppe every time we want to re-bias.
* Alters PrettyPrintEnv to keep unsuffixified and unaltered versions of each name alongside the altered version. This allows us to bias names properly even after suffixification
* Adds a `biasTo` combinator to PrettyPrintEnv which lets you re-sort the names for refs according to bias targets. This can be overridden at any point and doesn't require building a new PPE to do so.
* I ran into some crashes due to partial implementations of `Path.toName(')` so I added `Maybe` versions and denoted the others as unsafe, then migrated old usages where practical.
* Added `addFallback` to PrettyPrintEnv which allows explicitly providing a PPE to fallback on if-and-only-if a definition is missing from the first one.
  * This explicit fall-back behaviour provides the desired behaviour when viewing definitions by hash in Share, since you'll want to see all the names for a hash within the current scope, but want to fall back to global names for that hash iff there are none in your current scope.
* Removes Monoid/Semigroup from PrettyPrintEnv since there are now two reasonable semigroup instances and it's better to make callers be explicit about which they want.
* External names returned by the name lookup index were actually "relative", this changes them to be absolute as expected.

## Loose Ends

Biasing things *correctly* is still a bit unintuitive and is harder than I'd like, this is largely because the names inside a PPE might be absolute or relative, and generally you'll need to bias towards _both_ names in order to get the desired effect.

E.g. if the query is `map` and we're in `base.trunk.List`, we bias towards both `map` and `.base.trunk.List.map` which ensures we render the correct `map`, but also still prefer names in `trunk` over those in other releases even though they're outside of our perspective.

* Now that I'm keeping both suffixified and unsuffixified versions of each name in the ppe's result list it should be easy to remove the concept of a PrettyPrintEnvDecl entirely and just allow you to select the `suffixified` or `unsuffixified` version of a name from a single shared PPE. That can be done later in a separate PR, since there's still some potential for breakage there.

### CODE REVIEW

File: parser-typechecker/src/Unison/Codebase/Path.hs [hunk 1]
```diff
@@ -256,12 +258,24 @@ fromName' n = case take 1 (Name.toString n) of
     path = fromName n
     seq = toSeq path
 
-toName :: Path -> Name
-toName = Name.unsafeFromText . toText
+unsafeToName :: Path -> Name
```
Comment (ChrisPenner) [inline parser-typechecker/src/Unison/Codebase/Path.hs:261]: This function was partial but didn't indicate as such and caused me a few crashes while developing this, I marked it as partial and added a new safe `Maybe` version.
I also went through and patched some versions to the safe version wherever it was easy to do so.

File: parser-typechecker/src/Unison/PrettyPrintEnv.hs [hunk 2]
```diff
@@ -20,18 +28,25 @@ import qualified Unison.HashQualified' as HQ'
...
-    terms :: Referent -> Maybe (HQ'.HashQualified Name),
-    -- names for types
-    types :: Reference -> Maybe (HQ'.HashQualified Name)
+  { -- names for terms, constructors, and requests; e.g. [(original name, relativized and/or suffixified pretty name)]
+    termNames :: Referent -> [(HQ'.HashQualified Name, HQ'.HashQualified Name)],
+    -- names for types; e.g. [(original name, relativized and/or suffixified pretty name)]
+    typeNames :: Reference -> [(HQ'.HashQualified Name, HQ'.HashQualified Name)]
   }
 
+terms :: PrettyPrintEnv -> Referent -> Maybe (HQ'.HashQualified Name)
```
Comment (ChrisPenner) [inline parser-typechecker/src/Unison/PrettyPrintEnv.hs:44]: We re-export versions of the old fields for back-compatibility; these select the current 'top-priority' name for a reference.

File: unison-core/src/Unison/NamesWithHistory.hs [hunk 3]
```diff
@@ -250,33 +249,6 @@ termName length r NamesWithHistory {..} =
     hq n = HQ'.take length (HQ'.fromNamedReferent n r)
     isConflicted n = R.manyDom n (Names.terms currentNames)
 
-suffixedTypeName :: Int -> Reference -> NamesWithHistory -> [HQ'.HashQualified Name]
```
Comment (ChrisPenner) [inline unison-core/src/Unison/NamesWithHistory.hs:253]: This was deconstructed and moved to the `PrettyPrinterEnv.Names` module so we can share a prioritization scheme across both suffixified and unsuffixified versions.

File: parser-typechecker/src/Unison/Codebase/Path.hs [hunk 4]
```diff
@@ -399,10 +413,6 @@ instance Convert Path [NameSegment] where convert = toList
 
 instance Convert HQSplit (HQ'.HashQualified Path) where convert = unsplitHQ
 
-instance Convert Path Name where convert = toName
-
-instance Convert Path' Name where convert = toName'
-
```
Comment (aryairani) [inline parser-typechecker/src/Unison/Codebase/Path.hs:405]: These can become `Parse` instances instead, if that's useful.

File: parser-typechecker/src/Unison/PrettyPrintEnv.hs [hunk 5]
```diff
@@ -40,12 +56,36 @@ patterns ppe r =
...
+-- Whereas, a bias applied to the result of `addFallback` will bias within the available names
+-- inside the left PPE and will only search in the fallback if there aren't ANY names in the
+-- primary ppe.
+union :: PrettyPrintEnv -> PrettyPrintEnv -> PrettyPrintEnv
+union e1 e2 =
   PrettyPrintEnv
-    (\r -> terms e1 r <|> terms e2 r)
-    (\r -> types e1 r <|> types e2 r)
+    (\r -> termNames e1 r <|> termNames e2 r)
+    (\r -> typeNames e1 r <|> typeNames e2 r)
```
Comment (aryairani) [inline parser-typechecker/src/Unison/PrettyPrintEnv.hs]: `++` or at least `<>` would be clearer here than `<|>` (which was okay when they were `Maybe`).
```suggestion
    (\r -> termNames e1 r ++ termNames e2 r)
    (\r -> typeNames e1 r ++ typeNames e2 r)
```

File: parser-typechecker/src/Unison/PrettyPrintEnv.hs [hunk 6]
```diff
@@ -40,12 +56,36 @@ patterns ppe r =
...
+              else primaryNames
+    )
+
+-- | Finds names from both PPEs.
+--
+-- This is distinct from `addFallback` with respect to biasing;
+-- A bias applied to a union might select a name in the right half of the union.
+-- Whereas, a bias applied to the result of `addFallback` will bias within the available names
+-- inside the left PPE and will only search in the fallback if there aren't ANY names in the
+-- primary ppe.
```
Comment (aryairani) [inline parser-typechecker/src/Unison/PrettyPrintEnv.hs]: I would still call out that this is left-biased even though it's no longer called `unionLeft`; or consider leaving the old name.

File: parser-typechecker/src/Unison/PrettyPrintEnv.hs [hunk 7]
```diff
@@ -40,12 +56,36 @@ patterns ppe r =
 instance Show PrettyPrintEnv where
   show _ = "PrettyPrintEnv"
 
--- Left-biased union of environments
-unionLeft :: PrettyPrintEnv -> PrettyPrintEnv -> PrettyPrintEnv
-unionLeft e1 e2 =
+-- | Attempts to find a name in primary ppe, falls back to backup ppe only if no names are
+-- found.
```
Comment (aryairani) [inline parser-typechecker/src/Unison/PrettyPrintEnv.hs]: As someone writing code with this module, in what kind of situation would I want this combinator vs union?

File: unison-share-api/src/Unison/Server/Backend.hs [hunk 8]
```diff
@@ -635,10 +637,12 @@ getCurrentParseNames scope root =
 --      then name foo.bar.baz becomes baz
 --           name cat.dog     becomes .cat.dog
 fixupNamesRelative :: Path.Absolute -> Names -> Names
-fixupNamesRelative root = Names.map fixName
+fixupNamesRelative root names =
```
Comment (aryairani) [inline unison-share-api/src/Unison/Server/Backend.hs:638]: Is the comment for this function still accurate/representative?

**Review by aryairani — APPROVED** [review]

File: unison-share-api/src/Unison/Server/Backend.hs (hunk 8 above)
Comment (ChrisPenner) [reply unison-share-api/src/Unison/Server/Backend.hs:638]: Yes

File: parser-typechecker/src/Unison/Codebase/Path.hs (hunk 4 above)
Comment (ChrisPenner) [reply parser-typechecker/src/Unison/Codebase/Path.hs:405]: I don't believe we ever need to use it like that right now, and generally I find using the explicit combinators more clear 😄

File: parser-typechecker/src/Unison/Codebase/SqliteCodebase/Operations.hs [hunk 9]
```diff
@@ -565,11 +565,11 @@ namesAtPath path = do
       allTypes = typesInPath <> typesOutsidePath
   let rootTerms = Rel.fromList allTerms
   let rootTypes = Rel.fromList allTypes
-  let absoluteRootNames = Names {terms = rootTerms, types = rootTypes}
-  let absoluteExternalNames = Names {terms = Rel.fromList termsOutsidePath, types = Rel.fromList typesOutsidePath}
+  let absoluteRootNames = Names.makeAbsolute $ Names {terms = rootTerms, types = rootTypes}
```
Comment (ChrisPenner) [inline parser-typechecker/src/Unison/Codebase/SqliteCodebase/Operations.hs:568]: In a future PR, investigate whether we actually want absoluteRootNames in parseNames at all.

File: parser-typechecker/src/Unison/PrettyPrintEnv.hs [hunk 10]
```diff
@@ -4,34 +4,50 @@ module Unison.PrettyPrintEnv
...
 import qualified Unison.Referent as Referent
 
 data PrettyPrintEnv = PrettyPrintEnv
-  { -- names for terms, constructors, and requests
-    terms :: Referent -> Maybe (HQ'.HashQualified Name),
-    -- names for types
-    types :: Reference -> Maybe (HQ'.HashQualified Name)
+  { -- names for terms, constructors, and requests; e.g. [(original name, relativized and/or suffixified pretty name)]
+    termNames :: Referent -> [(HQ'.HashQualified Name, HQ'.HashQualified Name)],
+    -- names for types; e.g. [(original name, relativized and/or suffixified pretty name)]
```
Comment (ChrisPenner) [inline parser-typechecker/src/Unison/PrettyPrintEnv.hs]:
```suggestion
    -- names for types; e.g. [(original name, possibly suffixified name)]
```

File: parser-typechecker/src/Unison/PrettyPrintEnv.hs [hunk 11]
```diff
@@ -75,9 +119,46 @@ patternName env r =
...
+-- | Prefer names which share a common prefix with any provided target.
+--
+-- Results are sorted according to the longest common prefix found against ANY target.
+biasTo :: [Name] -> PrettyPrintEnv -> PrettyPrintEnv
+biasTo targets PrettyPrintEnv {termNames, typeNames} =
+  PrettyPrintEnv
+    { termNames = \r ->
+        r
+          & termNames
+          & Debug.debug Debug.Sync "pre-biased"
```
Comment (ChrisPenner) [inline parser-typechecker/src/Unison/PrettyPrintEnv.hs]: Remove or use a different flag

File: unison-cli/src/Unison/Codebase/Editor/HandleInput.hs [hunk 12]
```diff
@@ -2906,7 +2906,7 @@ getHQTerms = \case
             & Path.resolve currentPath'
             & Path.unabsolute
             & Path.toName
-    pure $ R.lookupRan path (Branch.deepTerms root0)
+    pure $ fromMaybe mempty $ (R.lookupRan <$> path <*> pure (Branch.deepTerms root0))
```
Comment (ChrisPenner) [inline unison-cli/src/Unison/Codebase/Editor/HandleInput.hs]: Swap to `unsafeToName`

File: unison-share-api/src/Unison/Server/Backend.hs [hunk 13]
```diff
@@ -823,7 +827,18 @@ prettyDefinitionsBySuffixes path root renderWidth suffixifyBindings rt codebase
...
+  (parseNames, localNamesOnly, unbiasedPPE) <- scopedNamesForBranchHash codebase root path
+  -- Bias towards both relative and absolute path to queries,
+  -- This allows us to still bias towards definitions outside our perspective but within the
+  -- same tree;
+  -- e.g. if the query is `map` and we're in `base.trunk.List`,
+  -- we bias towards `map` and `.base.trunk.List.map` which ensures we still prefer names in
+  -- `trunk` over those in other releases.
+  let biases =
+        mapMaybe HQ.toName query
+          & foldMap (\name -> [name, Name.makeAbsolute . Name.fromSegments $ NonEmpty.appendr (Path.toList path) (Name.segments name)])
```
Comment (ChrisPenner) [inline unison-share-api/src/Unison/Server/Backend.hs]: Investigate exactly this is here.

File: unison-share-api/src/Unison/Server/Backend.hs [hunk 14]
```diff
@@ -823,7 +827,18 @@ prettyDefinitionsBySuffixes path root renderWidth suffixifyBindings rt codebase
...
+  Debug.debugM Debug.Sync "Biases" biases
```
Comment (ChrisPenner) [inline unison-share-api/src/Unison/Server/Backend.hs]: Wrong debug flag; can be removed.

**Pull request merged**